quantum/impl/quantum_read_write_spinlock_impl.h

```diff
@@ -18,213 +18,257 @@
+    return SpinLockUtil::isWriteLocked(_count);
 }
 
 inline
 int ReadWriteSpinLock::numReaders() const
 {
-    int num = _count.load(std::memory_order_acquire);
-    return num == -1 ? 0 : num;
+    return SpinLockUtil::numReaders(_count);
 }
 
 inline
-ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock) :
-    _spinlock(lock),
-    _ownsLock(true)
+int ReadWriteSpinLock::numPendingWriters() const
 {
-    _spinlock.lockRead();
+    return SpinLockUtil::numPendingWriters(_count);
 }
 
+//==============================================================================================
+//                                ReadWriteSpinLock::Guard
+//==============================================================================================
 inline
-ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock, ReadWriteSpinLock::TryToLock) :
+ReadWriteSpinLock::Guard::Guard(ReadWriteSpinLock& lock,
+                                LockTraits::AcquireRead) :
     _spinlock(lock),
-    _ownsLock(_spinlock.tryLockRead())
+    _ownsLock(true),
+    _isUpgraded(false)
 {
+    _spinlock.lockRead();
 }
 
 inline
-ReadWriteSpinLock::ReadGuard::~ReadGuard()
+ReadWriteSpinLock::Guard::Guard(ReadWriteSpinLock& lock,
+                                LockTraits::AcquireWrite) :
+    _spinlock(lock),
+    _ownsLock(true),
+    _isUpgraded(true)
 {
-    if (_ownsLock) {
-        _spinlock.unlockRead();
-    }
+    _spinlock.lockWrite();
 }
 
 inline
-void ReadWriteSpinLock::ReadGuard::lock()
+ReadWriteSpinLock::Guard::Guard(ReadWriteSpinLock& lock,
+                                LockTraits::AcquireRead,
+                                LockTraits::TryToLock) :
+    _spinlock(lock),
+    _ownsLock(_spinlock.tryLockRead()),
+    _isUpgraded(false)
 {
-    if (!_ownsLock) {
-        _spinlock.lockRead();
-        _ownsLock = true;
-    }
 }
 
 inline
-bool ReadWriteSpinLock::ReadGuard::tryLock()
+ReadWriteSpinLock::Guard::Guard(ReadWriteSpinLock& lock,
+                                LockTraits::AcquireWrite,
+                                LockTraits::TryToLock) :
+    _spinlock(lock),
+    _ownsLock(_spinlock.tryLockWrite()),
+    _isUpgraded(_ownsLock)
 {
-    if (!_ownsLock) {
-        _ownsLock = _spinlock.tryLockRead();
-    }
-    return _ownsLock;
 }
 
 inline
-bool ReadWriteSpinLock::ReadGuard::ownsLock() const
+ReadWriteSpinLock::Guard::Guard(ReadWriteSpinLock& lock,
+                                LockTraits::AdoptLock) :
+    _spinlock(lock),
+    _ownsLock(lock.isLocked()),
+    _isUpgraded(lock.isWriteLocked())
 {
-    return _ownsLock;
 }
 
 inline
-void ReadWriteSpinLock::ReadGuard::unlock()
+ReadWriteSpinLock::Guard::~Guard()
 {
-    if (_ownsLock) {
-        _spinlock.unlockRead();
+    if (ownsLock()) {
+        unlock();
     }
 }
 
 inline
-ReadWriteSpinLock::WriteGuard::WriteGuard(ReadWriteSpinLock& lock) :
-    _spinlock(lock),
-    _ownsLock(true)
+void ReadWriteSpinLock::Guard::lock()
```

Comment:
Same comment as on ReadWriteMutex::Guard; we need lockRead() & lockWrite() rather than lock().